### PR TITLE
apiservers: stratify versioned informer construction

### LINKED ISSUE #50661: apiserver panic on 1.7.3  --admission-control=NamespaceLifecycle
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
Trying to upgrade from 1.6.7 to 1.7.3, using hyperkube api-server with the same configuration hit this error:

```
2017-08-15 06:25:19.419583 I | proto: duplicate proto type registered: google.protobuf.Any
2017-08-15 06:25:19.419727 I | proto: duplicate proto type registered: google.protobuf.Duration
2017-08-15 06:25:19.419759 I | proto: duplicate proto type registered: google.protobuf.Timestamp
I0815 06:25:19.462371       1 server.go:112] Version: v1.7.3+coreos.0
I0815 06:25:19.462685       1 aws.go:806] Building AWS cloudprovider
I0815 06:25:19.462737       1 aws.go:769] Zone not specified in configuration file; querying AWS metadata service
I0815 06:25:19.603659       1 tags.go:76] AWS cloud filtering on ClusterID: mycluster
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x16653e6]

goroutine 1 [running]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission/plugin/namespace/lifecycle.(*lifecycle).SetExternalKubeInformerFactory(0xc42079dd40, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission/plugin/namespace/lifecycle/admission.go:203 +0x26
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission/initializer.pluginInitializer.Initialize(0x8fcc9a0, 0xc420775600, 0x0, 0x0, 0x8f7c760, 0xc420eb81e0, 0x8f86160, 0xc42079dd40)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission/initializer/initializer.go:49 +0x112
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission/initializer.(*pluginInitializer).Initialize(0xc420ce0d80, 0x8f86160, 0xc42079dd40)
	<autogenerated>:1 +0x70
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission.PluginInitializers.Initialize(0xc420ec22a0, 0x2, 0x2, 0x8f86160, 0xc42079dd40)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission/plugins.go:180 +0x5d
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission.(*PluginInitializers).Initialize(0xc420ec22c0, 0x8f86160, 0xc42079dd40)
	<autogenerated>:18 +0x73
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission.(*Plugins).InitPlugin(0xc4202aae10, 0xc420409580, 0x12, 0x0, 0x0, 0x8f7c620, 0xc420ec22c0, 0x413390, 0xc420ec22a0, 0xc420ec2260, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission/plugins.go:155 +0x20c
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission.(*Plugins).NewFromPlugins(0xc4202aae10, 0xc42043b5e0, 0x7, 0x7, 0x8f7c660, 0xc42079d200, 0x8f7c620, 0xc420ec22c0, 0x0, 0xc420cd8480, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission/plugins.go:129 +0x126
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options.(*AdmissionOptions).ApplyTo(0xc420cf5470, 0xc42088ca80, 0xc420ec2260, 0x2, 0x2, 0x8fc5300, 0xc420f60900)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/admission.go:84 +0x453
k8s.io/kubernetes/cmd/kube-apiserver/app.BuildGenericConfig(0xc420483c00, 0x0, 0x0, 0x0, 0x2, 0x1, 0x0, 0x1, 0xc420d39d60, 0xc420d7cbd0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/server.go:448 +0x1150
k8s.io/kubernetes/cmd/kube-apiserver/app.CreateKubeAPIServerConfig(0xc420483c00, 0x0, 0x0, 0x8f76ba0, 0xc42024d860, 0x0, 0x6d17640, 0x28, 0x6cf0bcc, 0x5, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/server.go:254 +0xa9
k8s.io/kubernetes/cmd/kube-apiserver/app.Run(0xc420483c00, 0xc4200560c0, 0xc420ce99f8, 0xc4200ccff0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/server.go:119 +0x143
main.NewKubeAPIServer.func1(0xc42014c660, 0xc420799140, 0x0, 0xc, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/hyperkube/kube-apiserver.go:36 +0x36
main.(*HyperKube).Run(0xc420204a20, 0xc4200100f0, 0xc, 0xc, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/hyperkube/hyperkube.go:177 +0x3ec
main.(*HyperKube).RunToExit(0xc420204a20, 0xc4200100e0, 0xe, 0xe)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/hyperkube/hyperkube.go:187 +0x59
main.main()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/hyperkube/main.go:47 +0x847
```

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:
start apiserver with `NamespaceLifecycle`, it panic:
```
--admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,ResourceQuota,DefaultTolerationSeconds
```

after removing `NamespaceLifecycle`, it works:
```
--admission-control=LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,ResourceQuota,DefaultTolerationSeconds
```

**Anything else we need to know?**:
The pod startup script:
/etc/kubernetes/manifests/kube-apiserver.yml
```yml
apiVersion: v1
kind: Pod
metadata:
  name: kube-apiserver
  namespace: kube-system
spec:
  hostNetwork: true
  containers:
  - name: kube-apiserver
    image: quay.io/coreos/hyperkube:v1.7.3_coreos.0 
    command:
    - /hyperkube
    - apiserver
    - --cloud-provider=aws
    - --insecure-bind-address=0.0.0.0
    - --insecure-port=8080
    - --anonymous-auth=false
    - --secure-port=0
    - --apiserver-count=3 
    - --etcd-servers=http://etcd-cluster:2379
    - --allow-privileged=true
    - --service-cluster-ip-range=10.123.0.0/16
    - --advertise-address=10.0.233.106
    - --admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,ResourceQuota,DefaultTolerationSeconds
    - --runtime-config=extensions/v1beta1=true,extensions/v1beta1/networkpolicies=true
    ports:
    - containerPort: 443
      hostPort: 443
      name: https
    - containerPort: 8080
      hostPort: 8080
      name: local

```

**Environment**:
- Kubernetes version (use `kubectl version`):
 docker image -> quay.io/coreos/hyperkube:v1.7.3_coreos.0

- Cloud provider or hardware configuration**: AWS

- OS ( /etc/os-release):
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1409.7.0
VERSION_ID=1409.7.0
BUILD_ID=2017-07-19-0005
PRETTY_NAME="Container Linux by CoreOS 1409.7.0 (Ladybug)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"

- Kernel (e.g. `uname -a`):
Linux ip-10-0-233-106.ec2.internal 4.11.11-coreos #1 SMP Tue Jul 18 23:06:59 UTC 2017 x86_64 Intel(R) Xeon(R) CPU E5-2676 v3 @ 2.40GHz GenuineIntel GNU/Linux
- Install tools:
 custom setup

### PR DESCRIPTION
**Pull request by sttts**
The versioned share informer factory has been part of the GenericApiServer config,
but its construction depended on other fields of that config (e.g. the loopback
client config). Hence, the order of changes to the config mattered.

This PR stratifies this by moving the SharedInformerFactory from the generic Config
to the CompleteConfig struct. Hence, it is only filled during completion when it is
guaranteed that the loopback client config is set.

While doing this, the CompletedConfig construction is made more type-safe again,
i.e. the use of SkipCompletion() is considereably reduced. This is archieved by
splitting the derived apiserver Configs into the GenericConfig and the ExtraConfig
part. Then the completion is structural again because CompleteConfig is again
of the same structure: generic CompletedConfig and local completed ExtraConfig.

Fixes #50661.

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 1]
```diff
@@ -379,12 +390,18 @@ func (c *Config) Complete() completedConfig {
...
+	return CompletedConfig{&completedConfig{
+		c,
+		informers,
+	}}
 }
 
 // SkipComplete provides a way to construct a server instance without config completion.
-func (c *Config) SkipComplete() completedConfig {
-	return completedConfig{c}
+func (c *Config) SkipComplete(informers informers.SharedInformerFactory) CompletedConfig {
```
Comment (dixudx) [inline staging/src/k8s.io/apiserver/pkg/server/config.go]: `return CompletedConfig{&completedConfig{c, informers}}` ?

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 2]
```diff
@@ -121,8 +122,6 @@ type Config struct {
 	// Will default to a value based on secure serving info and available ipv4 IPs.
 	ExternalAddress string
 
-	// SharedInformerFactory provides shared informers for resources
-	SharedInformerFactory informers.SharedInformerFactory
 	//===========================================================================
```
Comment (dixudx) [inline staging/src/k8s.io/apiserver/pkg/server/config.go]: These three lines should be deleted as well ?

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 3]
```diff
@@ -300,11 +299,23 @@ func (c *Config) ApplyClientCert(clientCAFile string) (*Config, error) {
 
 type completedConfig struct {
 	*Config
+
+	//===========================================================================
```
Comment (dixudx) [inline staging/src/k8s.io/apiserver/pkg/server/config.go]: Should delete the newline here? These three commented lines should belong to `*Config`, right?

**Review by dixudx — COMMENTED** [review]
The stratified data structure looks pretty good.

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: They belong to the next fields.

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: It's header for everything following.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 4]
```diff
@@ -379,12 +390,18 @@ func (c *Config) Complete() completedConfig {
 		c.RequestInfoResolver = NewRequestInfoResolver(c)
 	}
 
-	return completedConfig{c}
+	return CompletedConfig{&completedConfig{
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: :+1:

**Comment by p0lyn0mial** [comment]
I see that my changes caused a lot of trouble :( I can assure I was not aware of the weak nature of `GenericApiServer config`. Thanks for sharing this.

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: fixed

**Comment by sttts** [comment]
@p0lyn0mial don't worry, a review of my patch would be nice though :-)

File: federation/cmd/federation-apiserver/app/server.go [hunk 5]
```diff
@@ -189,6 +191,12 @@ func NonBlockingRun(s *options.ServerRunOptions, stopCh <-chan struct{}) error {
 	}
 	sharedInformers := informers.NewSharedInformerFactory(client, 10*time.Minute)
 
+	clientgoExternalClient, err := clientgoclientset.NewForConfig(genericConfig.LoopbackClientConfig)
```
Comment (deads2k) [inline federation/cmd/federation-apiserver/app/server.go]: This block indicates to me that this isn't something that most API servers care about setting for themselves.

File: pkg/master/master.go [hunk 6]
```diff
@@ -138,6 +137,21 @@ type Config struct {
 	MasterCount int
 }
 
+type Config struct {
+	GenericConfig *genericapiserver.Config
+	ExtraConfig
```
Comment (deads2k) [inline pkg/master/master.go]: I really dislike anonymous inclusion.  I'm ok with the extra layer, but I'm not a fan of doing this.  Is it required to make the `completedConfig` work as expected?

File: pkg/master/master.go [hunk 7]
```diff
@@ -138,6 +137,21 @@ type Config struct {
...
+}
+
+type completedConfig struct {
+	GenericConfig genericapiserver.CompletedConfig
+	*ExtraConfig
+}
+
+type CompletedConfig struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*completedConfig
```
Comment (deads2k) [inline pkg/master/master.go]: This is all so we can trick the golang type system into doing something useful for us?

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go [hunk 8]
```diff
@@ -175,26 +171,6 @@ func (s *SecureServingOptions) ApplyTo(c *server.Config) error {
 		c.SecureServingInfo.SNICerts[server.LoopbackClientServerNameOverride] = &tlsCert
 	}
 
-	// create shared informers, if not explicitly set use in cluster config.
-	// do not fail on an error, this allows an external API server to startup
-	// outside of a kube cluster.
-	var clientCfg *rest.Config
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go]: This appears to the crux of it here.  We had to add this block of ugly for an exceptional case, but now the exceptional case is leaking out to everyone.  Only the kube-apiserver (and federation if they're doing something a little crazy (they probably are)) needs to do this.  Maybe they don't get to use `ApplyTo`?

File: pkg/master/master.go (hunk 6 above)
Comment (sttts) [reply pkg/master/master.go]: No, it was only convenience on my side. Will make it explicit.

**Comment by sttts** [comment]
@deads2k: ptal
- added informer construction to `RecommendedOptions.ApplyTo`
- unfolded ExtraConfig (in second commit, will squash if you agree)

**Comment by deads2k** [comment]
@sttts can you split out the separation of GenericConfig and ExtraConfig with complete changes separate from the RecommendedOptions bit?  I agree with the GenericConfig, ExtraConfig, and complete

**Comment by sttts** [comment]
@deads2k did they split you wanted.ptal. The RecommendedConfig/Options case now has no extra versioned informer as before.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 9]
```diff
@@ -187,6 +185,13 @@ type Config struct {
 	PublicAddress net.IP
 }
 
+type RecommendedConfig struct {
+	Config
+
+	// SharedInformerFactory provides shared informers for resources
+	SharedInformerFactory informers.SharedInformerFactory
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/config.go]: Expand the godoc explaining that is normally set via RecommendedOptions.ApplyTo.

File: staging/src/k8s.io/apiserver/pkg/server/options/recommended.go [hunk 10]
```diff
@@ -55,26 +60,33 @@ func (o *RecommendedOptions) AddFlags(fs *pflag.FlagSet) {
...
+	if err := o.Audit.ApplyTo(&config.Config); err != nil {
 		return err
 	}
-	if err := o.Features.ApplyTo(config); err != nil {
+	if err := o.Features.ApplyTo(&config.Config); err != nil {
 		return err
 	}
 
+	// do convenience work for RecommendedOptions users
+	clientgoExternalClient, err := clientgoclientset.NewForConfig(config.LoopbackClientConfig)
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/options/recommended.go]: This is what @p0lyn0mial discovered was odd.  This should be an in-cluster-config, not a loopback.

**Comment by deads2k** [comment]
> @deads2k did they split you wanted.ptal. The RecommendedConfig/Options case now has no extra versioned informer as before.

Thanks, the first three commits look good.  Had a couple comments on the last.

File: staging/src/k8s.io/apiserver/pkg/server/options/recommended.go (hunk 10 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/recommended.go]: right, I remember having seen that in the old code. will fix.

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 9 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: done

**Comment by sttts** [comment]
@p0lyn0mial @deads2k I restructured the `SharedInformersFactory` construction and brought back the `InClusterConfig` logic. As this is now in `RecommendedConfig` we don't need the loopback part anymore. Instead for out-of-cluster operation there is a `-kubeconfig` flag. Moreover, it's trivial to disable the `CoreAPI` logic altogether and plug-in your own logic to create the informers.

**Comment by deads2k** [comment]
> @p0lyn0mial @deads2k I restructured the SharedInformersFactory construction and brought back the InClusterConfig logic. As this is now in RecommendedConfig we don't need the loopback part anymore. Instead for out-of-cluster operation there is a -kubeconfig flag. Moreover, it's trivial to disable the CoreAPI logic altogether and plug-in your own logic to create the informers.

I like the new shape.

/lgtm

**Comment by dixudx** [comment]
@sttts Failed due to below undefined elements,

```
908 09:34:23.156] # k8s.io/kubernetes/test/integration/apiserver
W0908 09:34:23.157] test/integration/apiserver/apiserver_test.go:47: masterConfig.EnableCoreControllers undefined (type *master.Config has no field or method EnableCoreControllers)
W0908 09:34:23.910] # k8s.io/kubernetes/test/integration/auth
W0908 09:34:23.910] test/integration/auth/rbac_test.go:78: getter.config.StorageFactory undefined (type *master.Config has no field or method StorageFactory)
W0908 09:34:24.492] # k8s.io/kubernetes/test/integration/etcd
W0908 09:34:24.493] test/integration/etcd/etcd_storage_path_test.go:693: kubeAPIServerConfig.APIResourceConfigSource undefined (type *master.Config has no field or method APIResourceConfigSource)
W0908 09:34:48.446] # k8s.io/kubernetes/test/integration/master
W0908 09:34:48.447] test/integration/master/master_test.go:551: cfg.ServiceIPRange undefined (type *master.Config has no field or method ServiceIPRange)
```

**Comment by sttts** [comment]
@deads2k ptal at the two added commits. Will squash when you are happy.

**Pull request merged**